code/atom.dm

```diff
@@ -949,6 +949,23 @@
 /atom/proc/set_density(var/newdensity)
 	src.density = HAS_ATOM_PROPERTY(src, PROP_ATOM_NEVER_DENSE) ? 0 : newdensity
 
+/atom/proc/set_opacity(var/newopacity)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if (newopacity == src.opacity)
+		return // Why even bother
+
+	var/oldopacity = src.opacity
+	src.opacity = newopacity
+
+	SEND_SIGNAL(src, COMSIG_ATOM_SET_OPACITY, oldopacity)
+
+	if (!isturf(src))
```

Comment:
Random thought at work, should check if src.loc is a turf here instead. Non direct children arent in turf.contents anyways and wont affect opacity.